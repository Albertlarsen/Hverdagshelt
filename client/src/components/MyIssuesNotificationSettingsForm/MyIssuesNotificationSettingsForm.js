// @flow

import React from 'react';
import {NotificationSettingsService, UserService} from "../../services";
import jwt from "jsonwebtoken";
import {FormGroup, Checkbox, Button, Grid} from "react-bootstrap";
import {IssueNotificationSetting, User} from "../../classTypes";

let notificationSettingsService = new NotificationSettingsService();
let userService = new UserService();

interface State {
    notifications: IssueNotificationSetting,
    createNew: boolean,
    registered: boolean,
    inProgress: boolean,
    completed: boolean
}

export class MyIssuesNotificationSettingsForm extends React.Component <State> {
    state = {
        notifications: new IssueNotificationSetting('', 0, 0, 0),
        createNew: false,
        registered: false,
        inProgress: false,
        completed: false
    };

    onChangeHandler = (e) => {
        return undefined;
    };

    onclickHandler = (e) => {
        let value = -1;
        let bool = false;
        let newSettings = null;

        if (e === 1) {
            if (this.state.registered) {
                value = 0;
                bool = false;
            } else {
                value = 1;
                bool = true;
            }
            newSettings = new IssueNotificationSetting(this.state.notifications.userMail, value, this.state.notifications.inProgress, this.state.notifications.completed);
            this.setState({
                notifications: newSettings,
                registered: bool
            });
        } else if (e === 2) {
            if (this.state.inProgress) {
                value = 0;
                bool = false;
            } else {
                value = 1;
                bool = true;
            }
            newSettings = new IssueNotificationSetting(this.state.notifications.userMail, this.state.notifications.registered, value, this.state.notifications.completed);

            this.setState({
                notifications: newSettings,
                inProgress: bool
            });
        } else if (e === 3) {
            if (this.state.completed) {
                value = 0;
                bool = false;
            } else {
                value = 1;
                bool = true;
            }
            newSettings = new IssueNotificationSetting(this.state.notifications.userMail, this.state.notifications.registered, this.state.notifications.inProgress, value);
            this.setState({
                notifications: newSettings,
                completed: bool
            });
        }
    };

    componentDidMount() {
        let notificationObj = undefined;
        let val = false;
        let reg = false;
        let inProg = false;
        let compl= false;

        notificationSettingsService.getIssueNotificationSettings()
            .then(resources => {
                notificationObj = resources[0];
                if (notificationObj === undefined) {
                    val = true;
                    /*
                    notificationObj = new IssueNotificationSetting(this.state.user.mail, 0, 0, 0)
                    * alle brukere skal få default notification settings ved registrering
                    */
                    notificationObj = {
                        registered: 0,
                        inProgress: 0,
                        completed: 0
                    }
                } else {
                    if (notificationObj.registered === 1) reg = true;
                    if (notificationObj.inProgress === 1) inProg = true;
                    if (notificationObj.completed === 1) compl = true;
                }
                this.setState({
                    notifications: notificationObj,
                    createNew: val,
                    registered: reg,
                    inProgress: inProg,
                    completed: compl
                });
            });
    }

    render() {
        return (
            <form>
                <Grid>
                    <h3>Varsle meg når mine innsendte saker er:</h3>

                    <FormGroup id={1}>
                        <Checkbox checked={(this.state.registered > 0)}
                                  onChange={() => {
                                      this.onChangeHandler(1)
                                  }}
                                  onClick={() => {
                                      this.onclickHandler(1)
                                  }}>Registrert</Checkbox>
                    </FormGroup>
                    <FormGroup id={2}>
                        <Checkbox checked={(this.state.inProgress > 0)}
                                  onChange={() => {
                                      this.onChangeHandler(2)
                                  }}
                                  onClick={() => {
                                      this.onclickHandler(2)
                                  }}>Under arbeid</Checkbox>
                    </FormGroup>
                    <FormGroup id={3}>
                        <Checkbox checked={(this.state.completed > 0)}
                                  onChange={() => {
                                      this.onChangeHandler(3)
                                  }}
                                  onClick={() => {
                                      this.onclickHandler(3)
                                  }}>Fullført</Checkbox>
                    </FormGroup>

                    <Button  onClick={() => {
                        this.save()
                    }}>Lagre</Button>
                </Grid>
            </form>
        );
    }

    save = () => {
        if (this.state.createNew) {
            notificationSettingsService.addIssueNotificationSettings(this.state.notifications)
                .catch(error => {
                    console.log('Error when adding: ' + 'n\ ' + 'error: ' + error.message)
                })
        } else {
            notificationSettingsService.updateIssueNotificationSettings(this.state.notifications)
                .catch(error => {
                    console.log('Error when updating: ' + 'n\ ' + 'error: ' + error.message)
                })
        }
    }
}