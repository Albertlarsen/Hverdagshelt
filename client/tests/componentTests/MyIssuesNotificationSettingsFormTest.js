// @flow

import {shallow} from "enzyme/build";
import * as React from "react";
import {MyIssuesNotificationSettingsForm} from "../../src/components/MyIssuesNotificationSettingsForm/MyIssuesNotificationSettingsForm";

describe('Test for MyIssuesNotificationSettingsForm component', () => {
    const wrapper = shallow(<MyIssuesNotificationSettingsForm />);


    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

});