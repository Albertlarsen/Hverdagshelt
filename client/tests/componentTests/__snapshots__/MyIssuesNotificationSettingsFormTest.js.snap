// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for MyIssuesNotificationSettingsForm component renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MyIssuesNotificationSettingsForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Grid
        bsClass="container"
        componentClass="div"
        fluid={false}
      >
        <h3>
          Varsle meg når mine innsendte saker er:
        </h3>
        <FormGroup
          bsClass="form-group"
          id={1}
        >
          <Checkbox
            bsClass="checkbox"
            checked={false}
            disabled={false}
            inline={false}
            onChange={[Function]}
            onClick={[Function]}
            title=""
          >
            Registrert
          </Checkbox>
        </FormGroup>
        <FormGroup
          bsClass="form-group"
          id={2}
        >
          <Checkbox
            bsClass="checkbox"
            checked={false}
            disabled={false}
            inline={false}
            onChange={[Function]}
            onClick={[Function]}
            title=""
          >
            Under arbeid
          </Checkbox>
        </FormGroup>
        <FormGroup
          bsClass="form-group"
          id={3}
        >
          <Checkbox
            bsClass="checkbox"
            checked={false}
            disabled={false}
            inline={false}
            onChange={[Function]}
            onClick={[Function]}
            title=""
          >
            Fullført
          </Checkbox>
        </FormGroup>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          disabled={false}
          onClick={[Function]}
        >
          Lagre
        </Button>
      </Grid>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bsClass": "container",
        "children": Array [
          <h3>
            Varsle meg når mine innsendte saker er:
          </h3>,
          <FormGroup
            bsClass="form-group"
            id={1}
          >
            <Checkbox
              bsClass="checkbox"
              checked={false}
              disabled={false}
              inline={false}
              onChange={[Function]}
              onClick={[Function]}
              title=""
            >
              Registrert
            </Checkbox>
          </FormGroup>,
          <FormGroup
            bsClass="form-group"
            id={2}
          >
            <Checkbox
              bsClass="checkbox"
              checked={false}
              disabled={false}
              inline={false}
              onChange={[Function]}
              onClick={[Function]}
              title=""
            >
              Under arbeid
            </Checkbox>
          </FormGroup>,
          <FormGroup
            bsClass="form-group"
            id={3}
          >
            <Checkbox
              bsClass="checkbox"
              checked={false}
              disabled={false}
              inline={false}
              onChange={[Function]}
              onClick={[Function]}
              title=""
            >
              Fullført
            </Checkbox>
          </FormGroup>,
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            onClick={[Function]}
          >
            Lagre
          </Button>,
        ],
        "componentClass": "div",
        "fluid": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Varsle meg når mine innsendte saker er:",
          },
          "ref": null,
          "rendered": "Varsle meg når mine innsendte saker er:",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": <Checkbox
              bsClass="checkbox"
              checked={false}
              disabled={false}
              inline={false}
              onChange={[Function]}
              onClick={[Function]}
              title=""
            >
              Registrert
            </Checkbox>,
            "id": 1,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "checkbox",
              "checked": false,
              "children": "Registrert",
              "disabled": false,
              "inline": false,
              "onChange": [Function],
              "onClick": [Function],
              "title": "",
            },
            "ref": null,
            "rendered": "Registrert",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": <Checkbox
              bsClass="checkbox"
              checked={false}
              disabled={false}
              inline={false}
              onChange={[Function]}
              onClick={[Function]}
              title=""
            >
              Under arbeid
            </Checkbox>,
            "id": 2,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "checkbox",
              "checked": false,
              "children": "Under arbeid",
              "disabled": false,
              "inline": false,
              "onChange": [Function],
              "onClick": [Function],
              "title": "",
            },
            "ref": null,
            "rendered": "Under arbeid",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": <Checkbox
              bsClass="checkbox"
              checked={false}
              disabled={false}
              inline={false}
              onChange={[Function]}
              onClick={[Function]}
              title=""
            >
              Fullført
            </Checkbox>,
            "id": 3,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "checkbox",
              "checked": false,
              "children": "Fullført",
              "disabled": false,
              "inline": false,
              "onChange": [Function],
              "onClick": [Function],
              "title": "",
            },
            "ref": null,
            "rendered": "Fullført",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": false,
            "block": false,
            "bsClass": "btn",
            "bsStyle": "default",
            "children": "Lagre",
            "disabled": false,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Lagre",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Grid
          bsClass="container"
          componentClass="div"
          fluid={false}
        >
          <h3>
            Varsle meg når mine innsendte saker er:
          </h3>
          <FormGroup
            bsClass="form-group"
            id={1}
          >
            <Checkbox
              bsClass="checkbox"
              checked={false}
              disabled={false}
              inline={false}
              onChange={[Function]}
              onClick={[Function]}
              title=""
            >
              Registrert
            </Checkbox>
          </FormGroup>
          <FormGroup
            bsClass="form-group"
            id={2}
          >
            <Checkbox
              bsClass="checkbox"
              checked={false}
              disabled={false}
              inline={false}
              onChange={[Function]}
              onClick={[Function]}
              title=""
            >
              Under arbeid
            </Checkbox>
          </FormGroup>
          <FormGroup
            bsClass="form-group"
            id={3}
          >
            <Checkbox
              bsClass="checkbox"
              checked={false}
              disabled={false}
              inline={false}
              onChange={[Function]}
              onClick={[Function]}
              title=""
            >
              Fullført
            </Checkbox>
          </FormGroup>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            onClick={[Function]}
          >
            Lagre
          </Button>
        </Grid>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bsClass": "container",
          "children": Array [
            <h3>
              Varsle meg når mine innsendte saker er:
            </h3>,
            <FormGroup
              bsClass="form-group"
              id={1}
            >
              <Checkbox
                bsClass="checkbox"
                checked={false}
                disabled={false}
                inline={false}
                onChange={[Function]}
                onClick={[Function]}
                title=""
              >
                Registrert
              </Checkbox>
            </FormGroup>,
            <FormGroup
              bsClass="form-group"
              id={2}
            >
              <Checkbox
                bsClass="checkbox"
                checked={false}
                disabled={false}
                inline={false}
                onChange={[Function]}
                onClick={[Function]}
                title=""
              >
                Under arbeid
              </Checkbox>
            </FormGroup>,
            <FormGroup
              bsClass="form-group"
              id={3}
            >
              <Checkbox
                bsClass="checkbox"
                checked={false}
                disabled={false}
                inline={false}
                onChange={[Function]}
                onClick={[Function]}
                title=""
              >
                Fullført
              </Checkbox>
            </FormGroup>,
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="default"
              disabled={false}
              onClick={[Function]}
            >
              Lagre
            </Button>,
          ],
          "componentClass": "div",
          "fluid": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Varsle meg når mine innsendte saker er:",
            },
            "ref": null,
            "rendered": "Varsle meg når mine innsendte saker er:",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": <Checkbox
                bsClass="checkbox"
                checked={false}
                disabled={false}
                inline={false}
                onChange={[Function]}
                onClick={[Function]}
                title=""
              >
                Registrert
              </Checkbox>,
              "id": 1,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "checkbox",
                "checked": false,
                "children": "Registrert",
                "disabled": false,
                "inline": false,
                "onChange": [Function],
                "onClick": [Function],
                "title": "",
              },
              "ref": null,
              "rendered": "Registrert",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": <Checkbox
                bsClass="checkbox"
                checked={false}
                disabled={false}
                inline={false}
                onChange={[Function]}
                onClick={[Function]}
                title=""
              >
                Under arbeid
              </Checkbox>,
              "id": 2,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "checkbox",
                "checked": false,
                "children": "Under arbeid",
                "disabled": false,
                "inline": false,
                "onChange": [Function],
                "onClick": [Function],
                "title": "",
              },
              "ref": null,
              "rendered": "Under arbeid",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": <Checkbox
                bsClass="checkbox"
                checked={false}
                disabled={false}
                inline={false}
                onChange={[Function]}
                onClick={[Function]}
                title=""
              >
                Fullført
              </Checkbox>,
              "id": 3,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "checkbox",
                "checked": false,
                "children": "Fullført",
                "disabled": false,
                "inline": false,
                "onChange": [Function],
                "onClick": [Function],
                "title": "",
              },
              "ref": null,
              "rendered": "Fullført",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "active": false,
              "block": false,
              "bsClass": "btn",
              "bsStyle": "default",
              "children": "Lagre",
              "disabled": false,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Lagre",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
